### ISSUE TITLE
Add config defaults to meta

### ISSUE THREAD

**Issue by sed-i** [canonical/operator#775]
Currently, `CharmMeta` contains metadata.yaml and actions.yaml.

Since `ConfigData` is built from `config-get`, there is no way for a charm to know what is the "factory" default for a given config option.

Since we do not have functionality to keep/revert to "last known good configuration" (#189, #649) it might be handy to revert the "factory" default.

Practical use case: https://github.com/canonical/prometheus-k8s-operator/pull/308

https://github.com/canonical/operator/blob/2644ae9368c6b9915967a1e6a117b17ebacffb86/ops/charm.py#L725-L728

https://github.com/canonical/operator/blob/2644ae9368c6b9915967a1e6a117b17ebacffb86/ops/model.py#L964-L965

**Comment by rwcarlsen** [comment]
How do you imagine "reverting" to another (default) config state?  Design-wise, charms are not "allowed" to change config.  So being able to access config defaults wouldn't allow you to actually change the config from within the charm anyway.  Do you mean to access default config for purposes of "disobeying" the config change communicated by juju and just using the default value in the charm for "invalid config" cases?

**Comment by jnsgruk** [comment]
Indeed, this seems odd, and enabling a pattern that allows charms to disobey admins feels bad...

**Comment by rwcarlsen** [comment]
Saving admins from folly of their own self-destruction :joy:

**Comment by sed-i** [comment]
> charms are not "allowed" to change config.

> a pattern that allows charms to disobey admins

Yes, agreed, this is exactly how I described it in [lp/1969521](https://bugs.launchpad.net/juju/+bug/1969521).

> just using the default value in the charm for "invalid config" cases?

Yes. In second thought, for https://github.com/canonical/prometheus-k8s-operator/pull/308 it's probably not a good idea, but perhaps in other use cases such a fallback (+ blocked status) would make sense. And if not for "disobeying" the admin then perhaps for emitting a warning.

**Comment by rwcarlsen** [comment]
Logging a warning/error can be done for invalid config without knowing the defaults though.  @sed-i - do you have any other reasons or use cases to motivate adding defaults access?

**Comment by sed-i** [comment]
No further use cases at the moment.
Completeness of the meta information (i.e. the defaults) available via OF perhaps.
Can always be added if the need arises.
Feel free to close this.

**Comment by rwcarlsen** [comment]
Appreciate the ideas/discussion.  Going to close for now.
